src/control/server/storage/bdev/provider.go

```diff
@@ -0,0 +1,184 @@
+	PrepareResponse struct{}
+
+	// FormatRequest defines the parameters for a Format operation.
+	FormatRequest struct {
+		Forwarded  bool
+		Class      storage.BdevClass
+		DeviceList []string
+	}
+
+	// DeviceFormatResponse contains device-specific Format operation results.
+	DeviceFormatResponse struct {
+		Formatted  bool
+		Error      error
+		Controller *storage.NvmeController
+	}
+
+	// DeviceFormatResponses is a map of device identifiers to device Format results.
+	DeviceFormatResponses map[string]*DeviceFormatResponse
+
+	// FormatResults contains the results of a Format operation.
+	FormatResponse struct {
+		DeviceResponses DeviceFormatResponses
+	}
+
+	// Backend defines a set of methods to be implemented by a Block Device backend.
+	Backend interface {
+		Init(shmID ...int) error
+		Reset() error
+		Prepare(hugePageCount int, targetUser string, pciWhitelist string) error
+		Scan() (storage.NvmeControllers, error)
+		Format(pciAddr string) (*storage.NvmeController, error)
+	}
+
+	// Provider encapsulates configuration and logic for interacting with a Block
+	// Device Backend.
+	Provider struct {
+		log     logging.Logger
+		backend Backend
+	}
+)
+
+// DefaultProvider returns an initialized *Provider suitable for use in production code.
+func DefaultProvider(log logging.Logger) *Provider {
+	return NewProvider(log, defaultBackend(log))
+}
+
+// NewProvider returns an initialized *Provider.
+func NewProvider(log logging.Logger, backend Backend) *Provider {
+	return &Provider{
+		log:     log,
+		backend: backend,
+	}
+}
+
+// Init performs any initialization steps required by the provider.
+func (p *Provider) Init(req InitRequest) error {
+	return p.backend.Init(req.SPDKShmID)
+}
+
+// Scan attempts to perform a scan to discover NVMe components in the system.
+func (p *Provider) Scan(req ScanRequest) (*ScanResponse, error) {
+	cs, err := p.backend.Scan()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ScanResponse{
+		Controllers: cs,
+	}, nil
+}
+
+// Prepare attempts to perform all actions necessary to make NVMe components available for
+// use by DAOS.
+func (p *Provider) Prepare(req PrepareRequest) (*PrepareResponse, error) {
+	// run reset first to ensure reallocation of hugepages
+	if err := p.backend.Reset(); err != nil {
+		return nil, errors.WithMessage(err, "SPDK setup reset")
+	}
+
+	res := &PrepareResponse{}
+	// if we're only resetting, just return before prep
+	if req.ResetOnly {
+		return res, nil
+	}
+
+	return res, errors.WithMessage(
+		p.backend.Prepare(req.HugePageCount, req.TargetUser, req.PCIWhitelist),
+		"SPDK setup",
+	)
+}
+
+// Format attempts to initialize NVMe devices for use by DAOS (NB: no-op for non-NVMe devices).
+func (p *Provider) Format(req FormatRequest) (*FormatResponse, error) {
+	if len(req.DeviceList) == 0 {
+		return nil, errors.New("empty DeviceList in FormatRequest")
+	}
+
+	// TODO: Kick off device formats in goroutines? Serially formatting a large
```

Comment:
haven't done this so far because I'm not sure simultaneous device formats are possible through SPDK